eZ/Publish/Core/Persistence/Cache/PersistenceLogger.php

```diff
@@ -30,57 +35,168 @@ class PersistenceLogger
-    public function logCall($method, array $arguments = array())
+    public function logCall(string $method, array $arguments = []): void
+    {
+        ++$this->stats['uncached'];
+        if (!$this->logCalls) {
+            return;
+        }
+
+        $this->collectCacheCallData(
+            $method,
+            $arguments,
+            \array_slice(
+                \debug_backtrace(DEBUG_BACKTRACE_IGNORE_ARGS, 7),
+                2
+            ),
+            'uncached'
+        );
+    }
+
+    /**
+     * Log Cache miss, gets info it needs by backtrace if needed.
+     *
+     * @since 7.5
+     *
+     * @param array $arguments
+     * @param int $traceOffset
+     */
+    public function logCacheMiss(array $arguments = [], int $traceOffset = 2): void
+    {
+        ++$this->stats['miss'];
+        if (!$this->logCalls) {
+            return;
+        }
+
+        $trace = \debug_backtrace(DEBUG_BACKTRACE_IGNORE_ARGS, 7 + $traceOffset);
+        $this->collectCacheCallData(
+            $trace[$traceOffset - 1]['class'] . '::' . $trace[$traceOffset - 1]['function'],
+            $arguments,
+            \array_slice($trace, $traceOffset),
+            'miss'
+        );
+    }
+
+    /**
+     * Log a Cache hit, gets info it needs by backtrace if needed.
+     *
+     * @since 7.5
+     *
+     * @param array $arguments
+     * @param int $traceOffset
+     * @param bool $inMemory Denotes is cache hit was from memory (php variable), as opposed to from cache pool which
+     *                       is usually disk or remote cache service.
+     */
+    public function logCacheHit(array $arguments = [], int $traceOffset = 2, bool $inMemory = false): void
+    {
+        if ($inMemory) {
+            ++$this->stats['memory'];
+        } else {
+            ++$this->stats['hit'];
+        }
+
+        if (!$this->logCalls) {
+            return;
+        }
+
+        $trace = \debug_backtrace(DEBUG_BACKTRACE_IGNORE_ARGS, 7 + $traceOffset);
+        $this->collectCacheCallData(
+            $trace[$traceOffset - 1]['class'] . '::' . $trace[$traceOffset - 1]['function'],
+            $arguments,
+            \array_slice($trace, $traceOffset),
+            $inMemory ? 'memory' : 'hit'
+        );
+    }
+
+    /**
+     * Collection  method for {@see logCacheHit()}, {@see logCacheMiss()} & {@see logCall()}.
+     *
+     * @param $method
+     * @param array $arguments
+     * @param array $trimmedBacktrace
+     * @param string $type
+     */
+    private function collectCacheCallData($method, array $arguments, array $trimmedBacktrace, string $type): void
     {
-        ++$this->count;
-        if ($this->logCalls) {
-            $this->calls[] = array(
+        // simplest/fastests hash possible to identify if we have already collected this before to save on memory use
+        $callHash = \hash('adler32', $method . \serialize($arguments));
+        if (empty($this->calls[$callHash])) {
+            $this->calls[$callHash] = [
```

Comment:
As an improvement would be great have VO instead of arrays with fixed keys (also in terms of memory usage)

```php
final class PersistenceLoggerCall 
{
        /** @var string */
	private $method;
       /** @var array */
	private $arguments;
        /** @var \eZ\Publish\Core\Persistence\Cache\PersistenceLoggerStats */
	private $stats;
        /** @var \eZ\Publish\Core\Persistence\Cache\PersistenceLoggerCallTrace[] */
	private $traces;

	// ...
} 

final class PersistenceLoggerStats 
{
    /** @var int */
    private $uncached;
    /** @var int */
    private $miss;
    /** @var int */
    private $hit;
    /** @var int */
    private $memory;

    // ... 	
}  

final class PersistenceLoggerCallTrace 
{
    /** @var string[] */
    private $trace;
    /** @var int */   
    private $count;
}

```

Reply:
This looks like a followup if you want to work on it before final ;)